Задание 1. Базовое изучение
Изучить представленный набор данных на основе описания его столбцов и выбрать 8 столбцов для дальнейшего изучения (среди них должны быть как числовые, так и категориальные). Провести расчет базовых метрик для них, кратко описать результаты.

In [67]:
import pandas as pd

In [68]:
# Ознакомимся с таблицей
col_names = ['surgery', 'Age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'respiratory_rate', 'temperature_of_extremities', 'peripheral_pulse', 'mucous_membranes', 'capillary_refill_time', 'pain', 'peristalsis', 'abdominal_distension', 'nasogastric_tube', 'nasogastric_reflux', 'nasogastric_reflux_PH', 'rectal_examination_feces', 'abdomen', 'packed_cell_volume', 'total_protein', 'abdominocentesis_appearance', 'abdomcentesis_total_protein', 'outcome', 'surgical_lesion?', '25_type_of_lesion', '26_type_of_lesion', '27_type_of_lesion', 'cp_data']
df = pd.read_csv('horse_data.csv', header=None, na_values='?', usecols=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27], names=col_names)
df.head()

surgery  Age  Hospital_Number  rectal_temperature  pulse  respiratory_rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   

   temperature_of_extremities  peripheral_pulse  mucous_membranes  \
0                         3.0               3.0               NaN   
1                         NaN               NaN               4.0   
2                         1.0               1.0               3.0   
3                         4.0               1.0               6.0   
4                         NaN               NaN               6.0   

   capillary_refill_time  ...  packed_cell_volume  total_protein  \
0                    2.0  ...                45.0            8.4   
1                    1.0  ...                50.0           85.0   
2                    1.0  ...                33.0            6.7   
3                    2.0  ...                48.0            7.2   
4                    2.0  ...                74.0            7.4   

   abdominocentesis_appearance  abdomcentesis_total_protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   

   surgical_lesion?  25_type_of_lesion  26_type_of_lesion  27_type_of_lesion  \
0                 2              11300                  0                  0   
1                 2               2208                  0                  0   
2                 2                  0                  0                  0   
3                 1               2208                  0                  0   
4                 2               4300                  0                  0   

   cp_data  
0        2  
1        2  
2        1  
3        1  
4        2  

[5 rows x 28 columns]

In [69]:
# Посмотрим на статистические метрики
df.describe()

surgery         Age  Hospital_Number  rectal_temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory_rate  temperature_of_extremities  \
count  276.000000        242.000000                  244.000000   
mean    71.913043         30.417355                    2.348361   
std     28.630557         17.642231                    1.045054   
min     30.000000          8.000000                    1.000000   
25%     48.000000         18.500000                    1.000000   
50%     64.000000         24.500000                    3.000000   
75%     88.000000         36.000000                    3.000000   
max    184.000000         96.000000                    4.000000   

       peripheral_pulse  mucous_membranes  capillary_refill_time  ...  \
count        231.000000        253.000000             268.000000  ...   
mean           2.017316          2.853755               1.305970  ...   
std            1.042428          1.620294               0.477629  ...   
min            1.000000          1.000000               1.000000  ...   
25%            1.000000          1.000000               1.000000  ...   
50%            2.000000          3.000000               1.000000  ...   
75%            3.000000          4.000000               2.000000  ...   
max            4.000000          6.000000               3.000000  ...   

       packed_cell_volume  total_protein  abdominocentesis_appearance  \
count          271.000000     267.000000                   135.000000   
mean            46.295203      24.456929                     2.037037   
std             10.419335      27.475009                     0.804905   
min             23.000000       3.300000                     1.000000   
25%             38.000000       6.500000                     1.000000   
50%             45.000000       7.500000                     2.000000   
75%             52.000000      57.000000                     3.000000   
max             75.000000      89.000000                     3.000000   

       abdomcentesis_total_protein     outcome  surgical_lesion?  \
count                   102.000000  299.000000        300.000000   
mean                      3.019608    1.551839          1.363333   
std                       1.968567    0.737187          0.481763   
min                       0.100000    1.000000          1.000000   
25%                       2.000000    1.000000          1.000000   
50%                       2.250000    1.000000          1.000000   
75%                       3.900000    2.000000          2.000000   
max                      10.100000    3.000000          2.000000   

       25_type_of_lesion  26_type_of_lesion  27_type_of_lesion     cp_data  
count         300.000000         300.000000         300.000000  300.000000  
mean         3657.880000          90.226667           7.363333    1.670000  
std          5399.513513         649.569234         127.536674    0.470998  
min             0.000000           0.000000           0.000000    1.000000  
25%          2111.750000           0.000000           0.000000    1.000000  
50%          2673.500000           0.000000           0.000000    2.000000  
75%          3209.000000           0.000000           0.000000    2.000000  
max         41110.000000        7111.000000        2209.000000    2.000000  

[8 rows x 28 columns]

In [70]:
# посмотрим % пропусков в данных
for col in df.columns:
    pct_missing = df[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
Age - 0.0%
Hospital_Number - 0.0%
rectal_temperature - 20.0%
pulse - 8.0%
respiratory_rate - 19.3%
temperature_of_extremities - 18.7%
peripheral_pulse - 23.0%
mucous_membranes - 15.7%
capillary_refill_time - 10.7%
pain - 18.3%
peristalsis - 14.7%
abdominal_distension - 18.7%
nasogastric_tube - 34.7%
nasogastric_reflux - 35.3%
nasogastric_reflux_PH - 82.3%
rectal_examination_feces - 34.0%
abdomen - 39.3%
packed_cell_volume - 9.7%
total_protein - 11.0%
abdominocentesis_appearance - 55.0%
abdomcentesis_total_protein - 66.0%
outcome - 0.3%
surgical_lesion? - 0.0%
25_type_of_lesion - 0.0%
26_type_of_lesion - 0.0%
27_type_of_lesion - 0.0%
cp_data - 0.0%


In [71]:
df.Age.value_counts() 
# 2:  Age
#           1 = Adult horse
#           2 = Young (< 6 months) - Можно предположить, что иммелось ввиду 9 месяцев, и тогда надо было указать 2

1    276
9     24
Name: Age, dtype: int64

In [72]:
#Заменим 9 на 2
df_horse.loc[(df['Age']) == 9] = 2
df_horse['Age'].describe()

count    300.000000
mean       1.080000
std        0.271746
min        1.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        2.000000
Name: Age, dtype: float64

In [73]:
# Выберем 8 столбцов для дальнейшего изучения
df_horse = df[['surgery', 'Age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'temperature_of_extremities', 'total_protein', 'outcome']]
df_horse

surgery  Age  Hospital_Number  rectal_temperature  pulse  \
0        2.0    1           530101                38.5   66.0   
1        1.0    1           534817                39.2   88.0   
2        2.0    1           530334                38.3   40.0   
3        1.0    9          5290409                39.1  164.0   
4        2.0    1           530255                37.3  104.0   
..       ...  ...              ...                 ...    ...   
295      1.0    1           533886                 NaN  120.0   
296      2.0    1           527702                37.2   72.0   
297      1.0    1           529386                37.5   72.0   
298      1.0    1           530612                36.5  100.0   
299      1.0    1           534618                37.2   40.0   

     temperature_of_extremities  total_protein  outcome  
0                           3.0            8.4      2.0  
1                           NaN           85.0      3.0  
2                           1.0            6.7      1.0  
3                           4.0            7.2      2.0  
4                           NaN            7.4      2.0  
..                          ...            ...      ...  
295                         4.0           65.0      3.0  
296                         3.0            NaN      3.0  
297                         4.0            6.8      2.0  
298                         3.0            6.0      1.0  
299                         NaN           62.0      3.0  

[300 rows x 8 columns]

In [74]:
# Рассчитаем статистические метрики для количественных величин
df_horse[['rectal_temperature', 'pulse', 'total_protein']].describe()

rectal_temperature       pulse  total_protein
count          240.000000  276.000000     267.000000
mean            38.167917   71.913043      24.456929
std              0.732289   28.630557      27.475009
min             35.400000   30.000000       3.300000
25%             37.800000   48.000000       6.500000
50%             38.200000   64.000000       7.500000
75%             38.500000   88.000000      57.000000
max             40.800000  184.000000      89.000000

Краткое описание результатов:

Ошибка в возрасте лошадей - можно признать некорретным вводом

Age - Большая часть- взрослые лошади, большинство были прооперированы, были повтороные обращения в больницу.

Surgery - медиана - 1 (большинство случае не требовали хирургического вмешательства), при этом третий квартиль и среднее у старых лошадок больше, т.е. их оперируют чаще

rectal_temperature - количественная величина, 20.0% пропусков - пропуски в этом столбце можно заменить медианным значением, т.к. размах по этой величине небольшой, а пропусков очень много.

Pulse - количественная величина, 8.0% пропусков - можно попытаться заполнить средним значением в зависимости от других показателей(Age, rectal_temperature, temperature_of_extremities,...) скорее всего имеются выбросы, максимальный пульс 184, при нормальном 30-40. Медиана 64, 75% выборки менее 88.

Temperature_of_extremities - категориальная величина, 18.7% пропусков - можно попытаться заполнить относительно rectal_temperature. Большинство лошадей испытывают шок.

Total protein - количественная величина, 11.0% пропусков. Большинство лошадей страдает от обезвожживания, но это может быть ошибкой ввода данных.

Outcome - конечный показатель, 0.3% пропуска - можно заполнить на основаниии всех остальных показателей.  Более половины лошадей выжило.

Задание 2. Работа с выбросами

В выбранных числовых столбцах найти выбросы, выдвинуть гипотезы об их причинах и проинтерпретировать результаты. Принять и обосновать решение о дальнейшей работе с ними.

In [75]:
# Рассмотрим выбросы по величине rectal_temperature

q1 = df_horse['rectal_temperature'].quantile(0.25)
q3 = df_horse['rectal_temperature'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
print(f'{lower_bound} нижняя граница')
print(f'{upper_bound} верхняя граница')

df_rectal = df_horse[df_horse['rectal_temperature']>upper_bound]
df_rectal.head()

36.74999999999999 нижняя граница
39.550000000000004 верхняя граница


surgery  Age  Hospital_Number  rectal_temperature  pulse  \
20      1.0    1           530157                39.9   72.0   
54      2.0    1           529461                40.3  114.0   
75      1.0    9           534092                39.7  100.0   
91      2.0    1           529461                40.3  114.0   
99      2.0    1           530002                39.6  108.0   

    temperature_of_extremities  total_protein  outcome  
20                         1.0            6.1      1.0  
54                         3.0            8.1      3.0  
75                         3.0           57.0      3.0  
91                         3.0            8.1      2.0  
99                         3.0            8.0      1.0

In [76]:
# Рассмотрим выбросы по величине pulse

q1 = df_horse['pulse'].quantile(0.25)
q3 = df_horse['pulse'].quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr)
upper_bound = q3 + (1.5 * iqr)
print(f'{lower_bound} нижняя граница')
print(f'{upper_bound} верхняя граница')

df_pulse = df_horse[df_horse['pulse']>upper_bound]
df_pulse.head()

-12.0 нижняя граница
148.0 верхняя граница


surgery  Age  Hospital_Number  rectal_temperature  pulse  \
3        1.0    9          5290409                39.1  164.0   
41       2.0    9          5288249                39.0  150.0   
55       1.0    9          5282839                38.6  160.0   
255      1.0    9          5294539                38.8  184.0   
275      1.0    9          5297159                38.8  150.0   

     temperature_of_extremities  total_protein  outcome  
3                           4.0            7.2      2.0  
41                          NaN            8.5      1.0  
55                          3.0            NaN      2.0  
255                         1.0            3.3      2.0  
275                         1.0            6.2      2.0

Краткий вывод:
rectal_temperature - не сильно отличаются от среднего, чтобы вызывать подозрение.

pulse - Данные значения пульса выбиваются из общей картины, среднем пульс должен быть 20-25, при этом третий квартиль у нас 88, Пульс 150-184 в 2-3 раза больше. Нужно обратиться к специалисту для интерпритации, т.к. либо опечатка, либо показания пулься были получены при пограничном состоянии здоровья лошади (повышенная температура может свидетельствовать о заболевании), либо был испорчен прибор.

Задание 3. Работа с пропусками

Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [108]:
# Процентное отображение пропусков
for col in df_horse.columns:
    pct_missing = df_horse[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.3%
Age - 0.0%
Hospital_Number - 0.0%
rectal_temperature - 20.0%
pulse - 8.0%
temperature_of_extremities - 18.7%
total_protein - 11.0%
outcome - 0.3%


In [130]:
#Заполняем пропуски медианным значением в зависимости от возраста и наличия операций
df['rectal_temperature'].fillna(df.groupby(['Age', 'surgery'])['rectal_temperature'].transform('median'), inplace=True)
df['temperature_of_extremities'].fillna(df.groupby(['Age', 'surgery'])['temperature_of_extremities'].transform('median'), inplace=True)
df['total_protein'].fillna(df.groupby(['Age', 'surgery'])['total_protein'].transform('median'), inplace=True)
df['pulse'].fillna(df.groupby(['Age', 'surgery'])['pulse'].transform('median'), inplace=True)
df_2 = df[['surgery', 'Age', 'Hospital_Number', 'rectal_temperature', 'pulse', 'temperature_of_extremities', 'total_protein', 'outcome']]
df_2.dropna(inplace=True)
df_2

C:\Users\89539\anaconda3\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


surgery  Age  Hospital_Number  rectal_temperature  pulse  \
10       1.0    1           528548                38.1   66.0   
54       2.0    1           529461                40.3  114.0   
64       1.0    1           529667                39.0   64.0   
91       2.0    1           529461                40.3  114.0   
104      1.0    1           527563                37.8   52.0   
189      1.0    1          5299603                38.3   60.0   

     temperature_of_extremities  total_protein  outcome  
10                          3.0            6.0      1.0  
54                          3.0            8.1      3.0  
64                          3.0            7.5      1.0  
91                          3.0            8.1      2.0  
104                         1.0            6.6      2.0  
189                         3.0            6.0      1.0

In [132]:
# Проверим на результат
for col in df_2.columns:
    pct_missing = df_2[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery - 0.0%
Age - 0.0%
Hospital_Number - 0.0%
rectal_temperature - 0.0%
pulse - 0.0%
temperature_of_extremities - 0.0%
total_protein - 0.0%
outcome - 0.0%


In [133]:
df_2[['rectal_temperature', 'pulse', 'total_protein']].describe()

rectal_temperature       pulse  total_protein
count            6.000000    6.000000       6.000000
mean            38.966667   78.333333       7.050000
std              1.105743   28.040447       0.981326
min             37.800000   52.000000       6.000000
25%             38.150000   61.000000       6.150000
50%             38.650000   65.000000       7.050000
75%             39.975000  102.000000       7.950000
max             40.300000  114.000000       8.100000